### DISCUSSION TITLE
Pi Pico W and utelnet connection

### DISCUSSION THREAD

**Discussion by ESPWarren** [micropython/micropython#14239] [RP2040 / Pico]
Hi

I have a Pi Pico W. It has utelnet.py as one of its modules. I can impoort the module, and start the telnet server.

I am accessing the Pico using Thonny on my laptop. I can connect to the telnet server on the Pico from the laptop by using the command line 'telnet' command. When I connect, I get the usual intro text. On Thonny the telnet server reports an incoming established connection.

But that is as far as it goes. Here is the output I see on my laptop:

Trying 192.168.2.68...
Connected to 192.168.2.68.
Escape character is '^]'.

After that point, nothing I type makes any difference. Nothing is output to the screen.

I get precisely the same when I connect using a Linux Putty client.

I feel sure it has something to do with a simple config, but it is a long time since I used utelnet...

Any ideas?

Many thanks

**Comment by sophiedegran** [comment]
Hi
That's not a problem of micropython .. please have a look at 
https://www.programmersought.com/article/13912854727/

**Reply by ESPWarren** [reply]
Thank you for a reply.

That article seems to be about 'how to exit the telnet connection'

My problem is once I have established a telnet connection to the Pico W, I cannot do anything ON the Pico. I cannot access the Pico interpreter command line (as in thonny)...

It establishes a telnet connection, but does not fully give me access to write code or commands into the Pico.

I think this is not a Telnet issue, bit something about using it with micropython on the Pico - no?

Thanks

**Reply by sophiedegran** [reply]
If you see' telnet> ' you are connected .. if not stop monitor with Thonny ...

**Reply by ESPWarren** [reply]
YES! I DO get a Telnet connection.

I DO see 'telnet>'. I am connected

My problem is "What then?" - why can I not interact with the Pico itself?

On an ESP8266 with utelnet, I could see the outpot of REPL.

But with the Pico I see nothing.

Is it actually possible to use utelnet to access the Pico?

**Comment by robert-hh** [comment]
I tried that here using a Pico W and the command line telnet app of Linux, and generally it works. It is not the fastest thing on the world, and starts to get really tardy if the USB connection to the board is closed. So there's something to be improved. 
Better use webrepl.

**Reply by ESPWarren** [reply]
Thank you Robert for those helpful comments.

Yes, I recall that utelnet was never that good!

Can you point me to a simple 'how to' for using webrepl with the Pico?

Thanks

**Reply by sophiedegran** [reply]
Super! you know robert that was the same thing about that 30 years before!

**Reply by sophiedegran** [reply]
Close as resolved..

**Reply by robert-hh** [reply]
I just made my way though it (again), and it's kind of bumpy. 
The good thing is: webrepl.py  and webrepl_setup.py exists on a Pico. You need the webrepl app for your browser. https://github.com/micropython/webrepl You have to install it on your PC.
There are two ways to use it:
a) Manual config:
```
start Wifi
import webrepl
webrepl.start(password="whatever you like")
```
Then open the webrepl app in the browser, set the IP address in the upper left corner and push reset. You should be asked for a password. Enter it, and then you should be connected.

b) Set up a configuration.
That's done with webrepl_setup.py, which also already exists on the board. webrepl_setup requires some code lines in boot.py, like:
```
#import webrepl 
#webrepl.start()
```
Otherwise it flags an error. If these lines do not exist, add them and then run:
`import webrepl_setup`
and enter the requested data, which is just enable/disable and the password.
If that's done, you can start webrepl with:
```
import webrepl
webrepl.start()
```
That will be added to your boot.py file.
Webrepl is not the most shiny app, and there was a discussion to include something like webrepl in mpremote.

BUT, webrepl has the same problem as telnet, when the USB connection is absent of getting slooooow. Probably because they rely both on dupterm, and the USB connection has to wait for a timeout if USB is not present. I faintly recall that this was discussed sometimes.

@sophiedegran Being old, I might remember 30 year old topics better than recent ones, so this may have been discussed within the recent years.

**Reply by robert-hh** [reply]
> Close as resolved..

Not yet, I assume.

**Reply by ESPWarren** [reply]
Actually, the general issue is not closed. It is about how to remotely access the Pico using something like telnet or webrepl.

To answer my own question, the way to use webrepl is here:

https://docs.micropython.org/en/latest/esp8266/tutorial/repl.html

There are THREE essential points to make this work:

1) The wifi network MUST be set up before issuing 'webrepl.start()' - and add a time.sleep(1) between the establishment of a wifi connection and starting webrepl

2) It will not work properly if you also have the USB connection active. Webrepl seems only to work as expected when the Pico W is standalone

3) You have to add the port number (8266) for the webrepl client to connect to. So if your Pico is at 192.168.2.68 then you whpuld make sure the address in the top left of the webrepl client is we://192.168.2.68:8266

Hope that helps.

You may now close this thread.

Thank you

**Reply by ESPWarren** [reply]
Sorry Robert - I crossed over with your post!

Just to say, I have the webrepl invocation in my main.py file. It works fine. But having an established network with stability seems critical before webrepl.start()

Thanks again

**Reply by robert-hh** [reply]
> The wifi network MUST be set up before issuing 'webrepl.start()'

That improves telnet response as well.
